hkmc2/shared/src/main/scala/hkmc2/bbml/ConstraintSolver.scala

```diff
@@ -117,12 +120,21 @@ class ConstraintSolver(infVarState: InfVarUid.State, tl: TraceLogger):
   private def constrainArgs(lhs: TypeArg, rhs: TypeArg)(using Ctx, CCtx, TL): Unit =
     constrainImpl(rhs.negPart, lhs.negPart)
     constrainImpl(lhs.posPart, rhs.posPart)
-  
+
+  private def inlineSkolemBounds(ty: Type, pol: Bool)(using cache: Set[Uid[InfVar]]): Type = ty match
+    case v @ InfVar(_, uid, state, skolem) if skolem && !cache(uid) =>
+      given Set[Uid[InfVar]] = cache + uid
+      inlineSkolemBounds(if pol then (v :: state.upperBounds).reduceLeft(_ & _) else (v :: state.lowerBounds).reduceLeft(_ | _), pol)
+    case ComposedType(lhs, rhs, p) => ComposedType(inlineSkolemBounds(lhs, pol), inlineSkolemBounds(rhs, pol), p)
+    case NegType(ty) => NegType(inlineSkolemBounds(ty, !pol))
+    case ty: CachedBasicType => inlineSkolemBounds(ty.mkBasic, pol)
+    case _: ClassType | _: FunType | _: InfVar | _: TypeExt | Top | Bot => ty
```

Comment:
The `TypeExt` case is wrong. You should call `toBasic` on `ty` so you don't have to handle this case.